server/sonar-server/src/main/java/org/sonar/server/measure/live/MeasureMatrix.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.measure.live;
+
+import com.google.common.collect.ArrayTable;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Table;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.OptionalDouble;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.sonar.api.measures.Metric;
+import org.sonar.core.util.stream.MoreCollectors;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.measure.LiveMeasureDto;
+import org.sonar.db.metric.MetricDto;
+import org.sonar.server.computation.task.projectanalysis.qualitymodel.Rating;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+class MeasureMatrix {
+
+  // component uuid -> metric key -> measure
+  private final Table<String, String, MeasureCell> table;
+
+  // direction is from file to project
+  private final List<ComponentDto> bottomUpComponents;
+
+  private final Map<String, MetricDto> metricsByKeys;
+
+  MeasureMatrix(List<ComponentDto> bottomUpComponents, Collection<MetricDto> metrics, List<LiveMeasureDto> dbMeasures) {
+    this.bottomUpComponents = bottomUpComponents;
+    this.metricsByKeys = metrics
+      .stream()
+      .collect(MoreCollectors.uniqueIndex(MetricDto::getKey));
+    this.table = ArrayTable.create(Lists.transform(bottomUpComponents, ComponentDto::uuid), metricsByKeys.keySet());
+    Map<Integer, MetricDto> metricsByIds = metricsByKeys.values()
+      .stream()
+      .collect(MoreCollectors.uniqueIndex(MetricDto::getId));
+    for (LiveMeasureDto dbMeasure : dbMeasures) {
+      table.put(dbMeasure.getComponentUuid(), metricsByIds.get(dbMeasure.getMetricId()).getKey(), new MeasureCell(dbMeasure, false));
+    }
+  }
+
+  ComponentDto getProject() {
+    return bottomUpComponents.get(bottomUpComponents.size() - 1);
+  }
+
+  Stream<ComponentDto> getBottomUpComponents() {
```

Comment:
any reason to expose a Stream instead of a List?

Reply:
no need to have an intermediary list. Used only a single place.